### PR TITLE
feat(`cheatcodes`): `getArtifactPathByCode` and `getArtifactPathByDeployedCode`

### LINKED ISSUE #7634: feat(`cheatcodes`): add a `vm.getArtifactPath` cheatcode that gets the artifact path for a given contract address
### Component

Forge

## Problem

At Sphinx, it's necessary for us to programmatically find the artifact that corresponds to each contract deployment in the user's Forge script. This is currently a complicated process that involves iterating over build info files and checking for a bytecode match. This has a couple significant drawbacks:

* Build info files are inherently large, so reading them from the file system is very slow.
* We've essentially re-implemented a version of Foundry's [`diff_score` function](https://github.com/foundry-rs/foundry/blob/0c961f742a0a567edf4618b9e6f12d286bc3a51c/crates/common/src/contracts.rs#L89) to match the bytecode in an artifact to the bytecode deployed in the user's script.

We could remove all of this logic if Foundry provides a cheatcode that returns an artifact path given a contract's address or bytecode.

## Solution

We propose a new cheatcode, `getArtifactPath`, with the following interface:
```sol
/// Gets the artifact path for the given contract address.
function getArtifactPath(
    address contract
) external view returns (string memory artifactPath);
```

Under the hood, this cheatcode would get the deployed bytecode for the given address, then use Foundry's [`diff_score` function](https://github.com/foundry-rs/foundry/blob/0c961f742a0a567edf4618b9e6f12d286bc3a51c/crates/common/src/contracts.rs#L89) to find the artifact that matches the deployed bytecode. (I'm not very familiar with Foundry's codebase, but I imagine the implementation would resemble that description).

We'd be happy to implement this cheatcode ourselves as long as we have the green light from the Foundry devs. cc @mattsse @klkvr 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Closes #7634 

## Solution

Adds new cheatcodes
`function getArtifactPathByCode(bytes calldata code) external returns (string memory path);` 
`function getArtifactPathByDeployedCode(bytes calldata deployedCode) external returns (string memory path);`

### CODE REVIEW

File: crates/cheatcodes/src/fs.rs [hunk 1]
```diff
@@ -250,6 +250,28 @@ impl Cheatcode for writeLineCall {
...
+        let Self { contractName: name } = self;
+        let path = state
+            .config
+            .available_artifacts
+            .as_ref()
+            .and_then(|artifacts| {
+                artifacts
+                    .iter()
+                    .find(|(id, _)| {
+                        let id_name = id.name.split('.').next().unwrap();
```
Comment (mattsse) [inline crates/cheatcodes/src/fs.rs:264]: why do we need this?
isn't this the contract name already?

**Review by mattsse — CHANGES_REQUESTED** [review]
makes sense, wdyt @grandizzy 

needs test

File: crates/cheatcodes/src/fs.rs (hunk 1 above)
Comment (yash-atreya) [reply crates/cheatcodes/src/fs.rs:264]: ArtifactId.name could have the form `Counter.0.8.23`

**Comment by grandizzy** [comment]
this is great to have but doesn't really solve #7634 which requests an address to be passed as param and then match the deployed bytecode for the given address > use bytecode_diff_score to identify the contract / it's path

**Comment by grandizzy** [comment]
re failing test - we're at a point when a new cheatcode added breaks it so seed should be updated, pls see https://github.com/foundry-rs/foundry/pull/8882/files#diff-d6c9e2f7cec6e686d5693fd6c6bd6400e4f617e1dea2695df29cab51549ba797R163

**Comment by yash-atreya** [comment]
> this is great to have but doesn't really solve #7634 which requests an address (or bytecode) to be passed as param and then match the deployed bytecode > use bytecode_diff_score to identify the contract / it's path (`ContractsByArtifact.find_by_code` kind of what we're doing for coverage)

+1. Added `vm.getArtifactPath(creationCode)`, removing `vm.getArtifactPath(contractName)`. Not sure how we can look up artifact path using deployed address? wdyt @grandizzy ?

File: crates/cheatcodes/spec/src/vm.rs [hunk 2]
```diff
@@ -1531,6 +1531,10 @@ interface Vm {
     #[cheatcode(group = Filesystem)]
     function writeLine(string calldata path, string calldata data) external;
 
+    /// Gets the artifact path from the creation code.
+    #[cheatcode(group = Filesystem)]
+    function getArtifactPath(bytes calldata creationCode) external view returns (string memory path);
```
Comment (mattsse) [inline crates/cheatcodes/spec/src/vm.rs:1536]: imo it's not obvious what this function accepts as argument. should we make this more explicit, like getArtifactPathByCode or smth @grandizzy ?

**Review by mattsse — APPROVED** [review]
lgtm, defer to @grandizzy on how to proceed.

File: crates/cheatcodes/spec/src/vm.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs:1536]: agree that `getArtifactPathByCode` is cleaner, also since for `getCode` we have `getCode(artifactPath)` and `getDeployedCode(artifactPath)` maybe we need to add a `getArtifactPathByDeployedCode` too? (where the only difference is that `ContractsByArtifact.find_by_deployed_code` instead `find_by_creation_code` is called)

Comment (yash-atreya) [reply crates/cheatcodes/spec/src/vm.rs:1536]: +1 to both. 

Makes sense to name them explicitly, also avoids signature conflicts.

File: crates/forge/tests/it/fuzz.rs [hunk 3]
```diff
@@ -160,7 +160,7 @@ async fn test_scrape_bytecode() {
     let filter = Filter::new(".*", ".*", ".*fuzz/FuzzScrapeBytecode.t.sol");
     let mut runner = TEST_DATA_DEFAULT.runner();
     runner.test_options.fuzz.runs = 2000;
-    runner.test_options.fuzz.seed = Some(U256::from(100u32));
+    runner.test_options.fuzz.seed = Some(U256::from(120u32));
```
Comment (DaniPopes) [inline crates/forge/tests/it/fuzz.rs:163]: Why is this needed?

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/forge/tests/it/fuzz.rs (hunk 3 above)
Comment (yash-atreya) [reply crates/forge/tests/it/fuzz.rs:163]: Re: https://github.com/foundry-rs/foundry/pull/8938#issuecomment-2368871161

cc @grandizzy

Comment (grandizzy) [reply crates/forge/tests/it/fuzz.rs:163]: when we fuzz we use test contract bytecode which changed with adding the new cheatcode so there are now different fuzzed inputs, had similar here: https://github.com/foundry-rs/foundry/pull/8882/files#diff-d6c9e2f7cec6e686d5693fd6c6bd6400e4f617e1dea2695df29cab51549ba797R163
ref https://github.com/foundry-rs/foundry/issues/8115

Comment (DaniPopes) [reply crates/forge/tests/it/fuzz.rs:163]: I don't see how changing an interface (Vm) change bytecode of an unrelated contract

Comment (grandizzy) [reply crates/forge/tests/it/fuzz.rs:163]: the issue can be reproduced in master by appending a line like `function writeToml() external;` in `testdata/cheats/Vm.sol` and `cargo test test_scrape_bytecode` will start failing. It can make it pass if bumping fuzz runs from 2000 to 5000. let me create a simpler repro to expose such

Comment (grandizzy) [reply crates/forge/tests/it/fuzz.rs:163]: I made some more tests with a simpler repro and figured out that happens because test has target contract in test contract, so they get inline, if they're split then issue is not reproducible and test pass without changing the seed, here's a draft PR https://github.com/foundry-rs/foundry/pull/8953
@DaniPopes

Comment (grandizzy) [reply crates/forge/tests/it/fuzz.rs]: this change is no longer needed, the flaky test was fixed

**Review by DaniPopes — APPROVED** [review]
lgtm
